torch/_prims/__init__.py

```diff
@@ -404,7 +405,7 @@ def _elementwise_meta(
     # Acquires the dtype
     dtype = None
     scalar_type = None
-    for arg in args:
+    for arg in args_:
```

Comment:
this doesn't look right, args_with_fixed_dtypes are excluded from dtype determination on purpose (they shouldn't affect output dtype. 

Let's step back, what is the error that you are getting when using `where`, and what test fails?

Reply:
The error is still in current master.
The code snippet:
```
import torch
import torch._refs as refs

a = torch.randint(0, 2, (10,), dtype=torch.bool)
b = refs.where(a, 0, 1)     # "torch.where" works fine
```
Produces the following error:
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Input In [5], in <cell line: 6>()
      4 a = torch.randint(0, 2, (10,), dtype=torch.bool)
      5 b = torch.where(a, 0, 1)
----> 6 b = refs.where(a, 0, 1)

File ~/git/pytorch-master/torch/_prims/wrappers.py:173, in out_wrapper.<locals>._fn(out, *args, **kwargs)
    171 @wraps(fn)
    172 def _fn(*args, out=None, **kwargs):
--> 173     result = fn(*args, **kwargs)
    174     if out is not None:
    175         assert isinstance(out, TensorLike)

File ~/git/pytorch-master/torch/_prims/wrappers.py:114, in elementwise_type_promotion_wrapper.__call__.<locals>._fn(*args, **kwargs)
    107 promoted_args = {
    108     x: _maybe_convert_to_dtype(bound.arguments[x], compute_dtype)
    109     for x in self.type_promoting_arg_names  # type: ignore[union-attr]
    110     if x in bound.arguments.keys()
    111 }
    112 bound.arguments.update(promoted_args)
--> 114 result = fn(**bound.arguments)
    116 # FIXME?: assumes result is a single tensor
    117 assert isinstance(result, TensorLike)

File ~/git/pytorch-master/torch/_refs/__init__.py:1211, in where(pred, a, b)
   1208 assert pred.dtype is torch.bool
   1210 pred, a, b = _maybe_broadcast(pred, a, b)
-> 1211 return prims.where(pred, a, b)

File ~/git/pytorch-master/torch/_ops.py:49, in OpOverload.__call__(self, *args, **kwargs)
     48 def __call__(self, *args, **kwargs):
---> 49     return self._op(*args, **kwargs or {})

File ~/git/pytorch-master/torch/_prims/__init__.py:353, in _make_prim.<locals>._autograd_impl(*args, **kwargs)
    351 def _autograd_impl(*args, **kwargs):
    352     flat_args, args_spec = tree_flatten((args, kwargs))
--> 353     return BackwardsNotSupported.apply(args_spec, *flat_args)

File ~/git/pytorch-master/torch/_prims/__init__.py:343, in _make_prim.<locals>.BackwardsNotSupported.forward(ctx, args_spec, *flat_args)
    341 g = torch._C._AutoDispatchBelowAutograd()
    342 try:
--> 343     return _prim(*args, **kwargs)
    344 finally:
    345     del g

File ~/git/pytorch-master/torch/_ops.py:49, in OpOverload.__call__(self, *args, **kwargs)
     48 def __call__(self, *args, **kwargs):
---> 49     return self._op(*args, **kwargs or {})

File ~/git/pytorch-master/torch/_prims/__init__.py:330, in _make_prim.<locals>._prim_impl(*args, **kwargs)
    326 def _prim_impl(*args, **kwargs):
    327     # always run the meta function because aten implementation will
    328     # typically accept more inputs (e.g., it will do promotion and
    329     # broadcasting) which we want to reject
--> 330     meta(*args, **kwargs)
    331     return impl_aten(*args, **kwargs)

File ~/git/pytorch-master/torch/_prims/__init__.py:1848, in _where_meta(pred, a, b)
   1844 def _where_meta(
   1845     pred: TensorLikeType, a: TensorLikeType, b: TensorLikeType
   1846 ) -> TensorLikeType:
-> 1848     return _elementwise_meta(
   1849         a,
   1850         b,
   1851         type_promotion=ELEMENTWISE_PRIM_TYPE_PROMOTION_KIND.DEFAULT,
   1852         args_with_fixed_dtypes=(pred,),
   1853     )

File ~/git/pytorch-master/torch/_prims/__init__.py:433, in _elementwise_meta(type_promotion, args_with_fixed_dtypes, *args)
    429 # NOTE: type promotion behavior here is mostly hidden from tests because
    430 # references will typically handle the type promotion properly even if this doesn't
    431 # (but getting it wrong will cause too many casts to be inserted in traces!)
    432 if device is not None:
--> 433     assert dtype is not None
    434     if type_promotion == ELEMENTWISE_PRIM_TYPE_PROMOTION_KIND.DEFAULT:
    435         dtype = dtype

AssertionError: 
```

Reply:
Yes, we should be using `scalar_type` when deciding `dtype`, it's an oversight that we aren't (and an oversight that there's no tests for `where` with 2 scalar arguments).